services/httpd/handler.go

```diff
@@ -834,6 +851,58 @@ func (h *Handler) serveExpvar(w http.ResponseWriter, r *http.Request) {
 	fmt.Fprintln(w, "\n}")
 }
 
+// serveDebugRequests will track requests for a period of time.
+func (h *Handler) serveDebugRequests(w http.ResponseWriter, r *http.Request) {
+	var d time.Duration
+	if s := r.URL.Query().Get("seconds"); s == "" {
+		d = DefaultDebugRequestsInterval
+	} else if seconds, err := strconv.ParseInt(s, 10, 64); err != nil {
+		h.httpError(w, err.Error(), http.StatusBadRequest)
+		return
+	} else {
+		d = time.Duration(seconds) * time.Second
+		if d > MaxDebugRequestsInterval {
+			h.httpError(w, fmt.Sprintf("exceeded maximum interval time: %s > %s",
+				influxql.FormatDuration(d),
+				influxql.FormatDuration(MaxDebugRequestsInterval)),
+				http.StatusBadRequest)
+			return
+		}
+	}
+
+	var closing <-chan bool
+	if notifier, ok := w.(http.CloseNotifier); ok {
+		closing = notifier.CloseNotify()
+	}
+
+	profile := h.requestTracker.TrackRequests()
+
+	timer := time.NewTimer(d)
+	select {
+	case <-timer.C:
+		profile.Stop()
+	case <-closing:
+		// Connection was closed early.
+		profile.Stop()
+		timer.Stop()
+		return
+	}
+
+	w.Header().Add("Connection", "close")
+	requests := make(map[string]*RequestStats, len(profile.Requests))
+	for req, st := range profile.Requests {
```

Comment:
Since the profile has been stopped, it shouldn't be modified any longer. The lock should only be needed while the profile is happening.